schedule/schedule.go

```diff
@@ -0,0 +1,165 @@
+	}
+
+	if sch.Shifts == nil || len(sch.Shifts) == 0 {
+		fmt.Println("No input shifts found. Starting from scratch.")
+		return nil
+	}
+
+	var previousStartDate time.Time
+	for i, shift := range sch.Shifts {
+		if err := shift.Validate(); err != nil {
+			return fmt.Errorf("schedule is invalid because shift %v is invalid: %v", i, err)
+		}
+
+		if !previousStartDate.IsZero() && previousStartDate.After(shift.StartDate) {
+			return fmt.Errorf("shifts out of order: shift time %v found after %v", previousStartDate, shift.StartDate)
+		}
+		previousStartDate = shift.StartDate
+
+		if shift == sch.LastShift() {
+			if shift.StopDate.IsZero() {
+				return fmt.Errorf("missing stop date on last shift")
+			}
+		} else {
+			// All other shifts should not have a stop date.
+			if !shift.StopDate.IsZero() {
+				return fmt.Errorf("stop date is only valid on the last shift. Found stop date %v on shift at index %v", shift.StopDate, i)
+			}
+		}
+	}
+
+	return nil // It's all good.
+}
+
+// A shift represents a span of time someone is on duty.
+type Shift struct {
+	// The start date, inclusive, of when the User goes on duty. Only year, month, and day field are relevant.
+	StartDate time.Time `json:"startDate,omitempty"`
+	User      string    `json:"user,omitempty"`
+
+	// UserOverride should be used when Users want to swap or alter shift assignments without disturbing the normal
+	// rotation cycle.
+	UserOverride string `json:"userOverride,omitempty"`
+
+	// StopDate is inclusive, and must only be used on the last Shift of a Schedule. For all other Shifts, the stop date
+	// is implied by the next shift's StartDate, and this value should remain the zero `time.Time` value.
+	StopDate time.Time `json:"stopDate,omitempty"`
+}
+
+// StopDateExclusive is used when an API uses (start, stop) date ranges as (inclusive, exclusive). May returns a zero
+// `time.Time` when stop is not set.
+func (sh *Shift) StopDateExclusive() time.Time {
+	if sh.StopDate.IsZero() {
+		return sh.StopDate // zero value is both inclusive and exclusive.
+	}
+	return sh.StopDate.Add(24 * time.Hour)
+}
+
+// StopDateExclusive can be used when an API uses (start, stop) date ranges as (inclusive, exclusive).
+func (sh *Shift) SetStopDateExclusive(excl time.Time) {
+	if excl.IsZero() {
+		sh.StopDate = excl // zero value is both inclusive and exclusive.
+	}
+	sh.StopDate = excl.Add(-24 * time.Hour)
+}
+
+func (sh *Shift) ClearStopDate() {
+	sh.StopDate = time.Time{}
+}
+
+func (sh *Shift) Validate() error {
+	if sh == nil {
+		return fmt.Errorf("shift cannot be nil")
+	}
+
+	if sh.User == "" {
+		return fmt.Errorf("user cannot be empty")
+	}
+
+	if sh.StartDate.IsZero() {
+		return fmt.Errorf("start date cannot be zero value")
+	}
+
+	if !sh.StopDate.IsZero() && sh.StartDate.After(sh.StopDate) {
+		return fmt.Errorf("start date must be before stop date")
+	}
+
+	return nil
+}
+
+// MarshalJSON returns timestamps in the `DateFormat` format.
+func (sh *Shift) MarshalJSON() ([]byte, error) {
```

Comment:
why do these JSON functions exist? Is that just something the YAML parser requires?

Reply:
These functions are hooks the native [`encoding/json` package](https://golang.org/pkg/encoding/json/#Marshal) provides to do custom JSON [un]marshaling. Since our `yaml` package first translates all objects to/from JSON before then converting that to YAML, this custom hook allows us to interject the actual object we want to marshal to JSON, rather than the one we work with internally.

So why do we want to use an intermediate object in the process? Because the internal structure uses `time.Time` values, which by default produce output in the `Mon Jan 2 15:04:05 -0700 MST 2006` format, instead of our preferred and readable `Mon 02 Jan 2006`. 

Said in a different way, the process looks like this:
1. Go `struct`
1. intermediate Go `struct`
1. JSON
1. YAML

(human readable/editable YAML)

Unmarshaling goes the same way just in reverse. FWIW we could swap out the `github.com/ghodss/yaml` package for the `github.com/go-yaml/yaml` package, which skips the JSON intermediate format. I originally chose `ghodss` because the extra JSON step was needed to properly [un]marshal Proto structs.

Reply:
> FWIW we could swap out the github.com/ghodss/yaml package for the github.com/go-yaml/yaml package, which skips the JSON intermediate format.

Just curious if there are any reasons _not_ to do this?

(Also, sorry for being late to the party on this PR).

Reply:
I don't think there is any particular benefit of one library over the other - it's already coded up with `ghodss/yaml`, which happens to be the same package as `kleat`, so I'm inclined to use just for the familiarity (and portability to JSON if we would ever need that, or if protos ever do come into the mix).